레스토랑 상세 페이지 구현하기

dongakim opened the pull request.

dongakim committed:
Home 테스트 추가

dongakim committed:
Home 구현

dongakim committed:
App 테스트 추가

dongakim committed:
App 구현

dongakim committed:
Header 테스트 추가

dongakim committed:
Header 구현

dongakim committed:
Header 적용

dongakim committed:
About 테스트 추가

dongakim committed:
About 구현

dongakim committed:
Restaurants 테스트 추가

dongakim committed:
Restaurants 구현

dongakim committed:
App 리팩토링

dongakim committed:
Home 테스트 추가

dongakim committed:
Home 구현

dongakim committed:
Home 리팩토링

dongakim committed:
App 테스트 추가

dongakim committed:
App 구현

dongakim committed:
App 테스트 수정

Kihyun92 commented on src/App.jsx:
```diff
@@ -1,32 +1,13 @@
-  });
-
   return (
     <div>
```
fragment로 감싸줘도 될것 같아요~

Kihyun92 commented on src/App.test.jsx:
```diff
@@ -6,29 +6,44 @@ import { useDispatch, useSelector } from 'react-redux';
+    expect(getByText('헤더')).toBeInTheDocument();
+  });
+
+  context('with path /', () => {
```
추후엔 어떨지 모르겠지만, 지금 같은 상황에서는 context로 분리하는게 아니라 describe로 처리하는것도 좋을 것 같아요~ 

context는 주로 `with <-> without`
`when log in <-> when log out`
처럼 `같은 맥락에서 별개의 사례일 때` context를 사용합니다. 혹은 it 만으로는 설명이 너무 길어지는 경우에 사용하기도 합니다.

context 사용법은 아래 문서와 아샬님의 문서를 참고해주세요~

- https://www.betterspecs.org/
- https://github.com/ahastudio/til/blob/main/blog/2018/12-08-given-when-then.md

Kihyun92 replied:
ㅎㅎ 나머지 테스트들을 추가해주셨네요! 지금은 적합한 것 같아요~ 좋습니다 :)

Kihyun92 reviewed (changes requested):
시작이 좋네요 👍  라우팅을 어떻게 처리하실지 기대됩니다 ㅎㅎ
전반적인 구조부터 잡아보고 계신 것 같아요 ㅎㅎ 좋습니다.
`Red-Green-Refactoring` 사이클도 잘 가져가고 계시는것 같네요 :)
커밋을 조금 더 자세하게 작성해보셔도 좋을 것 같습니다~ 결국 작업자 본인에게도 도움이 되고 동료들에게도 도움이 될거에요 ☺️

dongakim committed:
App 리팩토링

dongakim committed:
Home 테스트 수정

요구사항에 맞게 link의 첫 글자를 대문자로 수정

dongakim committed:
Home 구현

테스트에 맞게 구현 수정함

dongakim committed:
react-router-dom 설치

dongakim committed:
App 테스트 추가

라우트를 위해 react-router-dom를 추가하고
라우트 테스트를 작성하고 기존의 테스트를 react-router-dom 환경에 맞게
변경함

dongakim committed:
App 구현

react-router-dom를 활용해 라우트 기능 구현

dongakim committed:
App 테스트 리팩토링

중복되는 부분 추상화

dongakim committed:
About 수정

요구사항에 맞게 heading 수정

dongakim committed:
App 테스트 추가

restaurants path 테스트 추가

dongakim committed:
App 구현

restaurants route 추가

dongakim committed:
Home 수정

react-router-dom 적용

dongakim committed:
Home 테스트 수정

react-router-dom 적용

dongakim committed:
RestaurantsContainer 테스트 추가

출력되는 레스토랑이 링크인지를 검증하는 테스트를 추가함

dongakim committed:
RestaurantsContainer 구현

레스토랑이 링크가 되도록 react-router-dom를 사용해 구현

dongakim committed:
Restaurants 테스트 수정

RestaurantsContainer 테스트를 위해 MemoryRouter 추가

dongakim committed:
RestaurantsContainer 리팩토링

link 수정 및 비구조할당 적용

dongakim committed:
RestaurantContainer 테스트 추가

dongakim committed:
RestaurantContainer 구현

테스트를 통과할 정도로만 간단하게 구현

dongakim committed:
reducer 테스트 추가

레스토랑 정보 추가 리듀서 테스트 추가

dongakim committed:
reducer 구현

restaurant 상태 추가

dongakim committed:
reducer 리팩토링

action creator 생성 및 적용

dongakim committed:
api 테스트 추가

레스토랑 조회 api 테스트 추가

dongakim committed:
api 구현

레스토랑 조회 api 구현

dongakim committed:
restaurant fixture 구현

dongakim committed:
restaurant fixture 구현

dongakim committed:
actions 테스트 추가

fetchRestaurant mock api 추가 및 loadRestaurant 비동기 action 테스트 추가

dongakim committed:
actions 구현

loadRestaurant 추가

dongakim committed:
reducer 테스트 수정

restaurant 초기값 설정

dongakim committed:
reducer 구현

restaurant 초기값 설정

dongakim committed:
RestaurantContainer 테스트 추가

react-redux 및 react-router-dom 적용

dongakim committed:
RestaurantContainer 구현

dongakim committed:
App 테스트 추가

restaurant route 테스트 추가

dongakim committed:
App 구현

restaurant route 구현

dongakim committed:
NotFoundPage 테스트 추가

dongakim committed:
NotFoundPage 구현

dongakim committed:
App 테스트 추가

Not found 페이지 path 테스트 추가

dongakim committed:
App 구현

Not found 페이지 route 구현

dongakim committed:
webpack config 추가

dongakim committed:
html-webpack-plugin 설치 및 적용

웹팩 main.js 404 에러 해결을 위해 적용. 원인은 아직 모름.

참고)
-https://github.com/CodeSoom/week6-assignment-1/pull/24#issuecomment-717796939
-https://stackoverflow.com/questions/51051490/react-webpack-dev-server-main-js-not-found

Kihyun92 commented on src/reducer.js:
```diff
@@ -4,6 +4,13 @@ const initialState = {
   regions: [],
   categories: [],
   restaurants: [],
+  restaurant: {
+    id: null,
+    categoryId: null,
+    name: null,
+    address: null,
+    menuItems: [],
```
restaurant객체가 매우 커진 것 같아요. 이 경우에는 initialState로 관리하듯이 한번 더 추상화 해주면 어떨까요?

Kihyun92 reviewed (changes requested):
좋습니다! `red-green-refactoring` 사이클을 잘 가져가고 계신 것 같아요 👍 
지금처럼 차근차근 추가해보시면 좋을 것 같습니다 ☺️

이전에 배웠던 `Container / Presentation component` 와 같이 관심사에 맞게 처리하고 있는지 항상 고민해보시면서 개발하면 좋을 것 같아요~

도움이 필요하면 언제든 말씀해주세요~

dongakim committed:
reducer 리팩토링

restaurant state가 비대해졌기때문에 한번더 추상화함.

Kihyun92 commented on src/App.jsx:
```diff
@@ -1,32 +1,24 @@
+        <Route exact path="/" component={Home} />
+        <Route path="/about" component={About} />
+        <Route exact path="/restaurants" component={Restaurants} />
+        <Route path="/restaurants/:restaurantId" component={RestaurantContainer} />
```
```suggestion
        <Route path="/restaurants/:id" component={RestaurantContainer} />
```

id 라고만해도 명확할 것 같은데 어떨까요? ☺️

Kihyun92 commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,36 @@
+        {menuItems.map((menuItem) => (
+          <li key={menuItem.id}>
+            {menuItem.name}
+          </li>
```
구조 분해 할당을 적용하면 좋을 것 같아요~

Kihyun92 commented on src/RestaurantContainer.jsx:
```diff
@@ -0,0 +1,36 @@
+
+import { loadRestaurant } from './actions';
+
+export default function RestaurantContainer() {
+  const { restaurantId } = useParams();
+  const dispatch = useDispatch();
+  const { restaurant } = useSelector((state) => state);
+  const { name, address, menuItems } = restaurant;
+
+  useEffect(() => {
+    dispatch(loadRestaurant(restaurantId));
+  }, []);
```
상세 페이지를 잘 구현해주신 것 같아요~

관심사 측면에서 봤을때 `RestaurantContainer`에서 많은 일을 하고 있는 것 같아요~
이전에 고민해본 `Container / Presentation component`, 라우팅 등이 있을 것 같은데, 어떤식으로 관리하면 좋을지 고민해보시면 좋을 것 같습니다~

Kihyun92 commented on src/reducer.js:
```diff
@@ -1,9 +1,18 @@
 import { equal } from './utils';
 
+const restaurantInitialState = {
```
`defaultRestaurant`같이 default 값 같은 느낌은 어떤가요? 적절한 네이밍을 찾는게 항상 어렵네요 😅

Kihyun92 reviewed (changes requested):
전반적으로 잘 구현해보신 것 같습니다! 👍 관심사에 따라 컴포넌트를 어떻게 나눌지 고민해보시면 좋을 것 같아요~

dongakim committed:
RestaurantContainer 리팩토링

메뉴 아이템 구조분해 할당

dongakim committed:
App 리팩토링

param 명 변경

dongakim committed:
RestaurantContainer 수정

메뉴 아이템 구조분해할당명 수정
param명을 id로 바꾸게되면서 같은 변수명이 되지 않도록.

dongakim committed:
Restaurant 테스트 추가

dongakim committed:
Restaurant 구현

dongakim committed:
RestaurantContainer 리팩토링

뷰에 대한 관심사를 Restaurant 컴포넌트로 분리함

Kihyun92 reviewed (approved):
고생하셨습니다~ 전반적으로 TDD도 잘 가져가신 것 같고, 이전 과제에서 배웠던 것들도 잘 적용해주셨네요 :) 그리고 커밋 메세지를 잘 작성해주셔서 수월하게 리뷰 드릴 수 있었던 것 같습니다 👍 

이번 과제를 하면서 느낀점이나 아쉬웠던 부분, 더 해보고 싶은 부분이 있다면 한번 고민해보시면 좋겠습니다~
